icicle/tests/test_hash_api.cpp

```diff
@@ -0,0 +1,215 @@
+  auto input = std::make_unique<uint32_t[]>(nof_input_limbs);
+  auto output = std::make_unique<uint32_t[]>(nof_input_limbs);
+  // Randomize the input array
+  randomize(input.get(), nof_input_limbs);
+
+  auto config = default_hash_config();
+  // Create Keccak-256 hash object
+  auto keccak256 = create_keccak_256_hash(nof_input_limbs);
+  // Run single hash operation
+  ICICLE_CHECK(keccak256.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+TEST_F(HashApiTest, Keccak512)
+{
+  const uint64_t nof_input_limbs = 16; // Number of input limbs
+  // Create unique pointers for input and output arrays
+  auto input = std::make_unique<uint32_t[]>(nof_input_limbs);
+  auto output = std::make_unique<uint32_t[]>(nof_input_limbs);
+  // Randomize the input array
+  randomize(input.get(), nof_input_limbs);
+
+  auto config = default_hash_config();
+  // Create Keccak-512 hash object
+  auto keccak512 = create_keccak_512_hash(nof_input_limbs);
+  // Run single hash operation
+  ICICLE_CHECK(keccak512.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+TEST_F(HashApiTest, sha3_256)
+{
+  const uint64_t nof_input_limbs = 16; // Number of input limbs
+  // Create unique pointers for input and output arrays
+  auto input = std::make_unique<uint32_t[]>(nof_input_limbs);
+  auto output = std::make_unique<uint32_t[]>(nof_input_limbs);
+  // Randomize the input array
+  randomize(input.get(), nof_input_limbs);
+
+  auto config = default_hash_config();
+  // Create sha3-256 hash object
+  auto sha3_256 = create_sha3_256_hash(nof_input_limbs);
+  // Run single hash operation
+  ICICLE_CHECK(sha3_256.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+TEST_F(HashApiTest, sha3_512)
+{
+  const uint64_t nof_input_limbs = 16; // Number of input limbs
+  // Create unique pointers for input and output arrays
+  auto input = std::make_unique<uint32_t[]>(nof_input_limbs);
+  auto output = std::make_unique<uint32_t[]>(nof_input_limbs);
+  // Randomize the input array
+  randomize(input.get(), nof_input_limbs);
+
+  auto config = default_hash_config();
+  // Create sha3-512 hash object
+  auto sha3_512 = create_sha3_512_hash(nof_input_limbs);
+  // Run single hash operation
+  ICICLE_CHECK(sha3_512.hash_single(input.get(), output.get(), config));
+  // TODO: Verify output (e.g., check CPU against CUDA)
+}
+
+/****************************** Merkle **********************************/
+class HashSumBackend : public HashBackend
+{
+public:
+  HashSumBackend(uint64_t total_input_limbs, uint64_t total_output_limbs, uint64_t total_secondary_input_limbs = 0)
+      : HashBackend(total_input_limbs, total_output_limbs, total_secondary_input_limbs)
+  {
+  }
+
+  virtual eIcicleError hash_single(
+    const limb_t* input_limbs,
+    limb_t* output_limbs,
+    const HashConfig& config,
```

Comment:
If you take it in the constructor you couple your tree to hash only data on device or on host but why would you do that? I don't see an advantage.

Reply:
That's for the user to decide but yes you are right as long as we can reuse the data on GPU. Otherwise it doesn't help much and also sometimes the data is too large for GPU memory.